### PR TITLE
This adds permission based features for viewing user specific borrowed books

### PR DESCRIPTION
**Pull request by hamadhassan3**
# Overview of the work
This project is part of the 3rd week in python training program. It implements a website based on Django web framework.

# Major Changes
1) A view for user specific borrowed books
2) A group for Librarian added.
3) All users in Librarian group can view all borrowed books.
4) Updated the interface for author details screen

# Output

Author Details Updated

author

Librarian features

librarian

User specific borrowed books

userbooks

# Feedback Required
I need feedback on the quality and style of the code. Please also mention if a bad practice is being used in the code or if there is a way to improve any part.

### CODE REVIEW

File: catalog/migrations/0006_alter_bookinstance_options.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    dependencies = [
+        ('catalog', '0005_bookinstance_borrower'),
+    ]
+
+    operations = [
+        migrations.AlterModelOptions(
+            name='bookinstance',
+            options={'ordering': ['due_back'], 'permissions': (('can_mark_returned', 'Set book as returned'),)},
+        ),
+    ]
```
Comment (shahabejaz) [inline catalog/migrations/0006_alter_bookinstance_options.py]: You could have used single migration file to make changes in the db. It looks a little odd to see two migration files in a single PR.

File: catalog/views.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
 from django.shortcuts import render
 from .models import Book, BookInstance, Language, Genre, Author
 from django.views import generic
-
-
+from django.contrib.auth.mixins import LoginRequiredMixin
+from django.contrib.auth.mixins import PermissionRequiredMixin
 
 def index(request):
```
Comment (shahabejaz) [inline catalog/views.py]: Always try to use type hinting when defining functions and variables. That style helps other developers as well as you to understand code quickly later on.

File: catalog/views.py [hunk 3]
```diff
@@ -96,3 +96,28 @@ class AuthorDetailView(generic.DetailView):
...
+    """Displays all books to users with permission"""
+    model = BookInstance
+    template_name ='catalog/bookinstance_list_all_borrowed_librarian.html'
+    paginate_by = 10
+
+    # The user must have these permissions to access this functionality
+    permission_required = 'catalog.can_mark_returned'
+
+
+    def get_queryset(self):
```
Comment (shahabejaz) [inline catalog/views.py]: Type hint will be useful here.

File: catalog/views.py [hunk 4]
```diff
@@ -96,3 +96,28 @@ class AuthorDetailView(generic.DetailView):
...
     paginate_by = 1
 
+
+class LoanedBooksByUserListView(LoginRequiredMixin, generic.ListView):
+    """Generic class-based view listing books on loan to current user."""
+    model = BookInstance
+    template_name ='catalog/bookinstance_list_borrowed_user.html'
+    paginate_by = 10
+
+    def get_queryset(self):
```
Comment (shahabejaz) [inline catalog/views.py]: Type hint will be useful here.

File: catalog/views.py [hunk 5]
```diff
@@ -96,3 +96,28 @@ class AuthorDetailView(generic.DetailView):
...
+    paginate_by = 10
+
+    # The user must have these permissions to access this functionality
+    permission_required = 'catalog.can_mark_returned'
+
+
+    def get_queryset(self):
+        """There is no filter on user id so all books are fetched"""
+
+        return BookInstance.objects.filter(status__exact='o').order_by('due_back')
```
Comment (shahabejaz) [inline catalog/views.py]: New line missing at the end.

File: catalog/migrations/0006_alter_bookinstance_options.py (hunk 1 above)
Comment (hamadhassan3) [reply catalog/migrations/0006_alter_bookinstance_options.py]: I have updated and included a single migration file now

File: catalog/views.py (hunk 2 above)
Comment (hamadhassan3) [reply catalog/views.py]: This issue was resolved in a previous PR. To fix this, I have simply resolved the conflicts this PR had with main (the type hints were already added in main, so simply retained main's contents). Is this the correct way to do this?

File: catalog/views.py (hunk 3 above)
Comment (hamadhassan3) [reply catalog/views.py]: I have added a type hint for this function now

File: catalog/views.py (hunk 4 above)
Comment (hamadhassan3) [reply catalog/views.py]: I have added a type hint here

File: catalog/views.py (hunk 5 above)
Comment (hamadhassan3) [reply catalog/views.py]: Fixed now

File: catalog/views.py (hunk 2 above)
Comment (shahabejaz) [reply catalog/views.py]: Yes.

**Review by shahabejaz — APPROVED** [review]

**Pull request merged**